### PR TITLE
#82 repository interfaces

### LINKED ISSUE #82: Repository Refactor to use interfaces.
Each Repository should have a separate interface and implementation

interface should be in `HMS/Repository/ThingRepository.php`
implementations should be in `HMS/Repository/Doctrine/DoctrineThingRepository.php`

### PR DESCRIPTION
**Pull request by dpslwk**
This closes #82 by refactoring all the repositories to have interfaces and implementations

I've also added profile and permission repos
and refactored userRepo->create to ->save

### CODE REVIEW

**Comment by geeksareforlife** [comment]
The code looks fine, and I'm testing it now.

The save functionality troubles me though - it is effectively the same function except that the entity we pass is different - can we bring this into a trait or something?

**Comment by dpslwk** [comment]
We could have a trait but then we can’t type hint.

Matt

**Comment by geeksareforlife** [comment]
We could if we abstracted it out to a private function in the trait that the public save function calls, but that might be overkill for what is a two line function!

**Comment by dpslwk** [comment]
Yeah thats over kill

Matt

File: app/HMS/Repositories/Doctrine/DoctrineRoleRepository.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    public function findAll()
+    {
+        $entities = parent::findAll();
+        $roles = [];
+        foreach ($entities as $entity) {
+            $roles[$entity->getName()] = $entity;
+        }
+
+        return new ArrayCollection($roles);
```
Comment (dpslwk) [inline app/HMS/Repositories/Doctrine/DoctrineRoleRepository.php]: I think the change of array vs ArrayCollection should be a separate PR once we have chosen which we want. And this PR should be mergers as is.
